[IBC] Implement ICS-23 CommitmentProof verification for the SMT

h5law opened the pull request:
## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 29 Jun 23 20:46 UTC
This pull request includes various changes across multiple files. 

In the `README.md` file, a new section titled "ICS-23 Vector Commitments" was added under the "Components" section, along with a link to the ICS-23 specification. Additionally, the link to the ICS-24 specification was updated to point to the `ics24.md` file.

In the `go.sum` file, new dependencies were added for the libraries `github.com/cosmos/gogoproto` and `github.com/h5law/ics23/go`, with specific version tags. The existing dependency on `github.com/pokt-network/smt` was also updated to a newer version.

The `ics23.md` file provides implementation details of ICS-23 Vector Commitments, including the benefits of using the `cosmos/ics23` library and the changes made for Pocket's implementation. It also explains the proof verification process for membership and non-membership proofs, accompanied by flowchart diagrams. The implementation logic can be found in the file `proofs_ics23.go`.

A new file `proofs_ics23_test.go` was added to the `ibc/store` package, containing test functions for generating and verifying commitment proofs using the ICS23 library. The tests cover membership and non-membership proofs for a sparse Merkle tree.

The `go.mod` file shows various changes, including the addition of replace directives for modules `github.com/cosmos/ics23/go` and `github.com/regen-network/gocuke`, updates to module versions, and the addition of a new module `github.com/cosmos/gogoproto`.

The file `proofs_ics23.go` is a new addition to the `ibc/store` package, containing functions and constants for verifying membership and non-membership in a Sparse Merkle Tree.

Lastly, in the file where the `NextCode` constant is defined, it was updated to a new value and a new constant and error function were added.

Please review these changes in the pull request.
<!-- reviewpad:summarize:end -->

## Issue

Fixes #841 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Add proof generation and verification in accordance with `cosmos/ics23`
- Add SMT proof -> ics23 proof conversion

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [x] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [x] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

Olshansk commented on ibc/docs/ics23.md:
```diff
@@ -0,0 +1,78 @@
+1. If the `ActualValueHash` field in the `ExclusionProof` is the SMT's placeholder value, then use the placeholder value as the leaf node hash and follow steps 3-4
+2. If the `ActualValueHash` field is not the placeholder value, then use the `ActualPath` and `ActualValueHash` fields to generate the leaf node hash - do not hash these values before hashing the node as they are populated from the SMT proof's `NonMembershipLeafData` field and thus are already hashed
+3. Hash the leaf node hash with the `SideNodes` found in the `Path` field of the `ExclusionProof` to generate the root hash
+4. Compare the root hash with the one provided
```
because we're using `ActualValueHash`, we also verify that the root hash is the same as the one computed even though it's different from the value the user gives?

Can you clarify this point

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,159 @@
+)
+
+var (
+	// Custom SMT spec as the store does not hash values
```
I keep bringing this up because I'm still personally not 100% sure myself (but trust your judgment), but you're confident that we should keep hashing the other values even though we're not hashing these ones?

Lmk if this is better for an offline dicussion.

h5law replied:
We dont hash any other values. The other side nodes are already hashed. When we insert into the tree there are 3 steps:
1. hash the key - this gives us the path (we use the `PathHasher`)
  - The `PathHasher` can be customised with the option `WithPathHasher()` when creating the tree, this cannot be `nil`
2. hash the value - this gives us the valueHash (we use the `ValueHasher`)
  - The `ValueHasher` can be customised with the option `WithValueHasher()` if the value hasher is `nil` we do a `noop` function when hashing the value leaving us with the raw bytes
3. We then hash both of these using the `TreeHasher` this is required when we create the tree - this gives us the nodes digest
  - The `TreeHasher` is used as the `PathHasher` and `ValueHasher` if they are not provided
  - The `TreeHasher` cannot be nil

From this we can see that we dont want to prehash the value as we need the raw bytes but the node is still encoded and hashed to give the digest whether the value is or not. The side nodes are digests, they are already hashed. What we do with them is create a new inner node like this:
```
innerNode = []byte{[]byte{1}, leftChild.digest, rightChild.digest}
```
And then we hash that to give us the next digest.

Olshansk replied:
The hashing and nonhashing makes sense to me in the context of IBC and SMT in independence.

It's not a blocker for this PR, but my question/concern is moreso for the larger context of the blockchain to make sure that I'm personally not missing/misunderstanding something. For the rest of the persistence module, question is: `Should we hash the values for the other trees or not?`

@dylanlott do you have thoughts/opinion here?

h5law replied:
So in my ICS-24 PR I made the decision to not hash any values in all of the trees (except root tree as its created differently) this is just to make things easier during creation. But open to ideas here

Olshansk replied:
Interesting. Would like to discuss this more (as a team). 

Moving the thread to https://github.com/pokt-network/pocket/pull/847#discussion_r1245898305

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,159 @@
+			EmptyChild:      make([]byte, 32),
+			Hash:            ics23.HashOp_SHA256,
+		},
+		MaxDepth:                   256,
```
Cool. Should be enough leaves

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,159 @@
+	right                 // 1
+)
+
+var (
```
Should probably mention that this has a direct dependency on constants /values from `pokt-network/smf`

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,159 @@
+	}
+}
+
+// convertSMPToExistenceProof converts a SparseMerkleProof to an ics23
```
```suggestion
// convertSMPToExclusionProof converts a SparseMerkleProof to an ics23
```

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,159 @@
+		},
+		InnerSpec: &ics23.InnerSpec{
+			ChildOrder:      []int32{0, 1},
+			ChildSize:       32,
```
Move `32` into a local constant and reuse it everywhere you use `32`

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,159 @@
+	actualPath := path[:]
+	actualValue := defaultValue
+	if proof.NonMembershipLeafData != nil {
+		actualPath = proof.NonMembershipLeafData[1:33]
```
```suggestion
		actualPath = proof.NonMembershipLeafData[1:33] // len(prefix): len(prefix)+`newConstSize`
```

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,159 @@
+// getPathBit determines whether the hash of a node at a certain depth in the tree is the
+// left or the right child of its parent
+func getPathBit(data []byte, position int) position {
+	if int(data[position/8])&(1<<(8-1-uint(position)%8)) > 0 {
```
Can you provide an example of what this does so others avoid spending the time figuring out what you already have

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,159 @@
+	for i := 0; i < len(sideNodes); i++ {
+		var prefix, suffix []byte
+		prefix = append(prefix, innerPrefix...)
+		if getPathBit(path, len(sideNodes)-1-i) == left {
```
Can you add a comment on what (and why) we're updating the suffic on left and prefix on right?

Olshansk commented on ibc/store/proofs_ics23_test.go:
```diff
@@ -0,0 +1,178 @@
+		expected      error
+	}{
+		{
+			// Successfully generates a membership proof for a key stored
```
You should name these and use the name below. This way, when a test fails, the name shows up in terminal (or CI)

Olshansk commented on ibc/store/proofs_ics23_test.go:
```diff
@@ -0,0 +1,178 @@
+	testCases := []struct {
+		key           []byte
+		value         []byte
+		nonmembership bool
```
s/nonmembership/isNonMembership

Olshansk commented on ibc/store/proofs_ics23_test.go:
```diff
@@ -0,0 +1,178 @@
+		if tc.fails {
+			require.EqualError(t, err, tc.expected.Error())
+			require.Nil(t, proof)
+			return
```
s/return/continue

Olshansk commented on ibc/store/proofs_ics23_test.go:
```diff
@@ -0,0 +1,178 @@
+			valid = VerifyMembership(root, proof, tc.key, tc.value)
+		}
+
+		if tc.valid {
+			require.True(t, valid)
+		} else {
+			require.False(t, valid)
+		}
```
require.Equal(t, tc.valid, valid)

h5law commented on ibc/docs/ics23.md:
```diff
@@ -0,0 +1,78 @@
+
+### Custom SMT `ProofSpec`
+
+The `ProofSpec` type in `cosmos/ics23` is used to define the steps needed to verify a proof, what hashing functions should be used, any node prefixes, etc. This is then passed into the verification functions in order to verify a proof, instead of having to interact with the tree itself. This is useful as proofs must be verified on a light-client and as such being able to verify a proof without cr ...
```
No so remember the way that we construct a proof requires us to provide the hashes of the neighbouring nodes so that when we hash the key-value pair in conjunction with the hashes in the proof (as long as we get their left/right order correct) we can reconstruct the root hash. The tree itself does this method without going back "into" the tree. What the using the tree does is that it allows us to use the hashing functions the tree was set up with without providing them again.

So when we do this without the tree it means
1. We do not need to create a tree
2. We follow the exact same steps
3. We have to provide the hashing functions
4. In the "steps" we have to say the order of the hashes

If we look at how we define the order of the hashes its kinda simple:
```go
func convertSideNodesToSteps(sideNodes [][]byte, path []byte) []*ics23.InnerOp {
	steps := make([]*ics23.InnerOp, 0, len(sideNodes))
	for i := 0; i < len(sideNodes); i++ {
		var prefix, suffix []byte
		prefix = append(prefix, innerPrefix...)
		if getPathBit(path, len(sideNodes)-1-i) == left {
			suffix = make([]byte, 0, len(sideNodes[i]))
			suffix = append(suffix, sideNodes[i]...)
		} else {
			prefix = append(prefix, sideNodes[i]...)
		}
		op := &ics23.InnerOp{
			Hash:   ics23.HashOp_SHA256,
			Prefix: prefix,
			Suffix: suffix,
		}
		steps = append(steps, op)
	}
	return steps
```

Then when we hash using this step with either a prefix or a suffix we add the current hash into the missing space and we have an inner node! This is then hashed to produce the next current hash and so on until we have the root.

Olshansk replied:
Got it. So the tl;dr of the point you're making is that merkle proofs are verifiable without the tree. Correct?

h5law commented on ibc/docs/ics23.md:
```diff
@@ -0,0 +1,78 @@
+
+### Converting `SparseMerkleProof` to `CommitmentProof`
+
+In order to convert the proofs generated by the SMT into a serialisable proof used by `cosmos/ics23`, the `SideNodes` field of the `SparseMerkleProof` must be converted into a list of `InnerOp` types which define the order of the hashes. The order of the hashes is important as depending on whether the next hash is the left or right neighbour of the current hash, they will be hashed in a different ...
```
There were old discussions in celestia's repo on this. It is possible and we could do it. It would be a heft task though. A cool idea would be setting `k` on creation. But as stated in those previous discussions we would need to benchmark it appropriately.

I am thinking of creating a suite of benchmark tests in the SMT we could then use as a baseline for any changes moving forward

h5law replied:
Yeah I'd like to explore this, as I mentioned it was a topic of discussion in the original SMT repo and would be cool to find our own optimisation - potentially for different node stores we could provide a optimal `k` constant:
```go
const (
  SimpleMapStoreK = 2
  BadgerK = 4
)
```
This could be cool.

h5law commented:
🚨 I HAVE REWRITTEN HISTORY 🚨

@Olshansk I went through all the comments, and rebased off of main. The PR is now much smaller without all the unecissary files from other PRs.

Olshansk commented on ibc/docs/ics23.md:
```diff
@@ -0,0 +1,89 @@
+3. Hash the leaf node hash with the `SideNodes` found in the `Path` field of the `ExclusionProof` to generate the root hash
+4. Compare the root hash with the one provided
+   - As the non-membership proof uses the `ActualValueHash` field to generate the leaf node hash, the non-membership proof is actually proving membership of either a placeholder key or an unrelated key in the tree.
+   - This means that if the root hash computed and the one provided are equal then the key we were looking for was not in the tree. If they are not equal the proof is invalid and the key is in the tree.
```
SOrry for the back & forth, but I'm personally still struggling to understand this. I tried to reword it so it's easier for the reader:

Is this correct?
```suggestion
   - if `computedRooHash == providedRootHash`: `key` not in tree -> `Proof` is valid -> exclusion QED
   - if `computedRooHash != providedRootHash`: `key is in tree` -> `Proof` is invalid  -> exclusion QED
```

Olshansk commented on ibc/docs/ics23.md:
```diff
@@ -0,0 +1,78 @@
+# ICS-23 Vector Commitments <!-- omit in toc -->
```
@h5law Stupid question below.

> . Hash the leaf node hash with the `SideNodes` found in the `Path` field of the `ExclusionProof` to generate the root hash

This looks like a merkle proof, so do you know why we call it a vector commitment?

h5law replied:
This is a merkle proof but they call them vector commitments in ICS-23 its the name of the spec heres how they define it:
> A vector commitment is a construction that produces a constant-size, binding commitment to an indexed vector of elements and short membership and/or non-membership proofs for any indices & elements in the vector.

AFAIK no one is actually using vector commitments they are all using merkle proofs for ICS-23

Olshansk commented on ibc/docs/ics23.md:
```diff
@@ -0,0 +1,89 @@
+2. Hash the leaf with the `SideNodes` found in the `path` field of the `ExistenceProof` to generate the root hash
+3. Compare the root hash with the one provided and expect them to be identical
+
+Non-membership proofs are verified as follows:
```
@h5law I don't know why I'm struggling to understand this documentation, but I put together this mermaid diagram. Can you edit it based on your understanding?

I still feel like I have a gap in my knowledge somewhere.

```mermaid

---
title: Exclusion Proof 
---
flowchart TD
    I["Proof,Key"]
    NMD{"proof.NonMembershipLeafData == nil ?"}
    KP1["actualPath = sha256(key) \n actualValue = placeholderNode"]
    KP2["actualPath = ProvidedKeyHash \n actualValue = ProvidedValueHash"]
    Compare{"ComputedRootHash == ProvidedRootHash ?"}
    R1{"actualValue == placeholderNode ?\n\n Tree matches - exclusion \n requires empty node for path(key)"}
    R2{"actualValue == providedValueHash ?\n\n Tree does not match - exclusion \n requires provided value at the path given"}    
    EV["Exclusion Prove VALID"]
    EI["Exclusion Prove INVALID"]

    I --> NMD
    NMD -- Yes --> KP1
    NMD -- No --> KP2

    KP1 -- Key,Value --> Compare
    KP2 -- Key,Value --> Compare
    
    Compare -- Yes --> R1
    Compare -- No --> R2

    R1 -- Yes --> EV
    R1 -- No --> EI
    R2 -- Yes --> EV
    R2 -- No --> EI    
```

h5law replied:
Have updated the docs with an accurate diagram

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,177 @@
+	"github.com/pokt-network/smt"
+)
+
+// position refers to whether the node is either the left or right child of its parent
```
```suggestion
// position refers to whether the node is either the left or right child of its parent for the binary smt
```

Olshansk commented on ibc/store/proofs_ics23_test.go:
```diff
@@ -0,0 +1,180 @@
+			if tc.fails {
+				require.EqualError(t, err, tc.expected.Error())
+				require.Nil(t, proof)
+				return
```
Noticed this was resolved without a comment. 

I'm okay with the pushback but would like to understand why.


![Screenshot 2023-06-28 at 3 22 07 PM](https://github.com/pokt-network/pocket/assets/1892194/32a490b5-b8f8-4b64-84ca-9ff9f14c1b08)

h5law replied:
My bad this was actually an error from reverting out of the `t.Run()` way of things with the name. Now its back the return should be there.

Olshansk replied:
Got it. Makes more sense now.

Olshansk commented on ibc/store/proofs_ics23_test.go:
```diff
@@ -0,0 +1,180 @@
+	tree := smt.NewSparseMerkleTree(nodeStore, sha256.New(), smt.WithValueHasher(nil))
+	require.NotNil(t, tree)
+
+	// Set a value in the store
```
```suggestion
	// Prepare a tree with a predetermined set of key-value pairs
```

Olshansk commented on ibc/store/proofs_ics23_test.go:
```diff
@@ -0,0 +1,180 @@
+	tree := smt.NewSparseMerkleTree(nodeStore, sha256.New(), smt.WithValueHasher(nil))
+	require.NotNil(t, tree)
+
+	// Set a value in the store
```
```suggestion
	// Prepare a tree with a predetermined set of key-value pairs
```

Olshansk commented on ibc/store/proofs_ics23.go:
```diff
@@ -0,0 +1,177 @@
+	return steps
+}
+
+// getPathBit takes the hash of a key (the path) and a position (depth) and returns whether at
```
See the ref - `Ref: https://github.com/pokt-network/smt/blob/main/utils.go`

Yea, this is definitely non-trivial to implement...

This business logic is VERY closely coupled with our specific SMT implementation. I'd suggest the following:
1. exposing `GetPathBit` in the smt and reusing it here
2. Move the comments you added to the smt
3. Add an `isLeftChild` helper and use it here

h5law replied:
I have added a techdebt comment here as I have opened [smt#14](https://github.com/pokt-network/smt/issues/14) which will cover this. As I have just release `v0.6.0` for the SMT i think that its best to bundle these small changes into releases to save on constantly updating the `go.mod`. As the functionality will not change until this issue in in progress I think its alright to leave it until then.

LMK wdyt

Olshansk replied:
Going to push back on this.

I think https://github.com/pokt-network/smt/issues/14 is a 1+ month research and development project, whereas exposing the `GetPathBit` function is something we can do before EOD.

h5law replied:
@Olshansk https://github.com/pokt-network/smt/pull/15

h5law commented on ibc/docs/ics23.md:
```diff
@@ -0,0 +1,89 @@
+   - **IMPORTANT**: DO NOT hash these values before hashing the node as they are populated from the SMT proof's `NonMembershipLeafData` field and thus are already hashed
+3. Hash the leaf node hash with the `SideNodes` found in the `Path` field of the `ExclusionProof` to generate the root hash
+4. Compare the root hash with the one provided
+   - As the non-membership proof uses the `ActualValueHash` field to generate the leaf node hash, the non-membership proof is actually proving membership of either a placeholder key or an unrelated key in the tree.
```
Yeah my bad good spot!

h5law commented on ibc/docs/ics23.md:
```diff
@@ -0,0 +1,89 @@
+
+Non-membership proofs are verified as follows:
+
+1. If the `ActualValueHash` field in the `ExclusionProof` is the SMT's placeholder value, then use the placeholder value as the leaf node hash and skip to step 3 below
```
Not made this change but instead clarified the statement

Olshansk commented on ibc/docs/ics23.md:
````diff
@@ -0,0 +1,116 @@
+     - `key` is in tree -> `Proof` is invalid -> exclusion QED
+
+```mermaid
+flowchart TD
````
Appreciate the new diagram!

Olshansk reviewed (approved):
ICS23!!!!!
